matrix/common/s3_zarr_store.py

```diff
@@ -0,0 +1,188 @@
+    "compressor": zarr.storage.default_compressor,
+    "dtypes": {
+        "expression": "<f4",
+        "cell_id": "<U64",
+        "cell_metadata": "<f4"
+    },
+    "order": "C"
+}
+
+
+class S3ZarrStore:
+
+    def __init__(self, request_id: str):
+        self._request_id = request_id
+        self._results_bucket = os.environ['S3_RESULTS_BUCKET']
+        self._zarr_lock_table_name = os.environ['DYNAMO_LOCK_TABLE_NAME']
+        self._cells_per_chunk = ZARR_OUTPUT_CONFIG['cells_per_chunk']
+        self.dynamo_handler = DynamoHandler()
+        self.s3_file_system = s3fs.S3FileSystem(anon=False)
+        self.exp_df = None
+        self.qc_df = None
+
+    def write_from_pandas_dfs(self, exp_df: DataFrame, qc_df: DataFrame, num_rows: int):
+        """Write specified number of rows from matrix dataframes to s3 results bucket.
+
+        Input:
+            exp_df: (DataFrame) expression pandas dataframe from dss zarr store
+            qc_df: (DataFrame) qc values pandas dataframe from dss zarr store
+            num_rows: (int) number of rows to write from input dataframes
+        """
+        self.exp_df = exp_df
+        self.qc_df = qc_df
+
+        # Figure out which rows of the output table this filtered chunk will be assigned.
+        output_start_row_idx, output_end_row_idx = self._get_output_row_boundaries(num_rows)
+        # Based on that, determine which zarr chunks we need to write to
+        output_start_chunk_idx, output_end_chunk_idx = self._get_output_chunk_boundaries(
+            output_start_row_idx,
+            output_end_row_idx
+        )
+
+        # Now iterate through each chunk we're supposed to write to, and write the
+        # appropriate rows to each one.
+        written_rows = 0
+        for chunk_idx in range(output_start_chunk_idx, output_end_chunk_idx):
+            output_chunk_start = chunk_idx * self._cells_per_chunk
+
+            # Get the start and end rows in the filtered matrix that correspond to
+            # this chunk as well as the start and end rows in the chunk.
+            input_row_start = int(written_rows)
+            output_row_start = int(max(0, output_start_row_idx - output_chunk_start))
+            input_row_end = int(min(output_end_row_idx - output_start_row_idx,
+                                    input_row_start + self._cells_per_chunk - output_row_start))
+            output_row_end = int(output_row_start + input_row_end - input_row_start)
+            print(f"Writing {input_row_start}:{input_row_end} --> {output_row_start}:{output_row_end}")
+
+            input_bounds = (input_row_start, input_row_end)
+            output_bounds = (output_row_start, output_row_end)
+            self._write_row_data_to_results_chunk(chunk_idx, input_bounds, output_bounds)
+            row_count = input_row_end - input_row_start
+            written_rows += row_count
+
+    def _write_row_data_to_results_chunk(self, chunk_idx: int, input_bounds: tuple, output_bounds: tuple):
```

Comment:
Is this tested?

Reply:
Left comment here to add test here.  Lower priority for this week. Codecov tool for all future PRs is a great idea so that we keep marching in the right direction